WordPress/gutenberg
# Editor: Make sure `resetEditorBlocks` is synchronous.

Fixes #21804

## Description

#21467 introduced a resolver for `getEditedEntityRecord` which made `resetEditorBlocks` asynchronous. This sometimes makes previous calls finish executing after newer ones, essentially undoing the editor to a stale block tree.

I added comments to the relevant code to explain:

```js
export function* resetEditorBlocks( blocks, options = {} ) {
	const {
		__unstableShouldCreateUndoLevel,
		selectionStart,
		selectionEnd,
	} = options;
	const edits = { blocks, selectionStart, selectionEnd };
	// First Run: `[ core/template-part: [] ], undefined`.
	// Second Run: `[ core/template-part: [ ...initialBlocks ] ], false`.
	console.log( blocks, __unstableShouldCreateUndoLevel );
	if ( __unstableShouldCreateUndoLevel !== false ) {
		const { id, type } = yield select( STORE_KEY, 'getCurrentPost' );
		const noChange =
			( yield select(
				'core',
				'getEditedEntityRecord',
				'postType',
				type,
				id
			) ).blocks === edits.blocks;
		// First Run: `false`.
		// Second Run: Doesn't reach this code.
		// getEditedEntityRecord having a resolver makes
		// the execution from the first run reach this point
		// after the second run finishes. This overwrites the
		// blocks with a stale value. A stale value without
		// template part inner blocks which causes the issue
		// described here.
		console.log( noChange );
		if ( noChange ) {
			return yield dispatch(
				'core',
				'__unstableCreateUndoLevel',
				'postType',
				type,
				id
			);
		}
		// We create a new function here on every persistent edit
		// to make sure the edit makes the post dirty and creates
		// a new undo level.
		edits.content = ( { blocks: blocksForSerialization = [] } ) =>
			serializeBlocks( blocksForSerialization );
	}
	yield* editPost( edits );
}
```

This PR fixes the issue by using a synchronous select data control like we used to have before.

But, I think that this problem could surface elsewhere, so the long-term solution should be more general. We should think about canceling stale action execution. That also raises the questions of how to define and continue with any cleanup operations, though.

## How to test this?

Verify that #21804 is no longer an issue; that you can insert existing template parts on the full site editing experiment as expected.

## Types of Changes

*Bug Fix:* `resetEditorBlocks` will now execute sequentially, avoiding an issue where older executions would overwrite newer ones.
*New Feature:* `@wordpress/data-controls` now has a synchronous version of `select`, `syncSelect`, which does not wait for resolvers.


## Checklist:
